compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -1605,131 +1606,114 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         let cause = ObligationCause::misc(span, block.owner.def_id);
 
         let mut fulfillment_errors = Vec::new();
-        let mut applicable_candidates: Vec<_> = infcx.probe(|_| {
-            // Regions are not considered during selection.
-            let self_ty = self_ty
-                .fold_with(&mut BoundVarEraser { tcx, universe: infcx.create_next_universe() });
-
-            struct BoundVarEraser<'tcx> {
-                tcx: TyCtxt<'tcx>,
-                universe: ty::UniverseIndex,
-            }
-
-            // FIXME(non_lifetime_binders): Don't assign the same universe to each placeholder.
-            impl<'tcx> TypeFolder<TyCtxt<'tcx>> for BoundVarEraser<'tcx> {
-                fn interner(&self) -> TyCtxt<'tcx> {
-                    self.tcx
-                }
+        infcx.probe(|_| {
+            let mut universes = if self_ty.has_escaping_bound_vars() {
+                let mut max_visitor =
```

Comment:
I think you can use `outer_exclusive_binder` fn here.

Like: `vec![None; self_ty.outer_exclusive_binder().as_usize()]`